### PR TITLE
Deploy only one celery instance

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: <link> N/A

### Overview

Run celery only in one unit. To achieve that, the celery container is removed and it will be planned instead on a single unit along with the indico process

### Rationale

Celery does not provide exactly once guarantees, so multiple units can cause tasks to be executed by multiple workers, impacting user experience when those have a visible result.

### Juju Events Changes

* leader elected
* peer relation departed

### Module Changes

charm.py

### Library Changes

N/A

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -457,9 +469,8 @@ def _get_indico_secret_key_from_relation(self) -> Optional[str]:
             secret_value = peer_relation.data[self.app].get("secret-key")
```
Comment (yanksyoon) [inline src/charm.py:469]: Out of the scope of this PR but out of curiosity, may I ask what secret this is referring to? (Perhaps we can start naming secrets as well)

File: pyproject.toml [hunk 2]
```diff
@@ -16,7 +16,7 @@ line-length = 99
 target-version = ["py38"]
 
 [tool.coverage.report]
-fail_under = 97
+fail_under = 96
```
Comment (yanksyoon) [inline pyproject.toml:19]: I'm guessing the drop is from the pebble layer changes?

**Review by yanksyoon — DISMISSED** [review]
LGTM! Just questions out of curiousity!

File: pyproject.toml (hunk 2 above)
Comment (arturo-seijas) [reply pyproject.toml]: Oops, I've reverted that change

**Review by yanksyoon — APPROVED** [review]

**Review by mthaddon — COMMENTED** [review]
If I'm understanding this correctly we're relying on the `peer-relation-departed` event happening before `leader-elected`. If it doesn't, we'd remove the nominated unit but not nominate a new unit until another `leader-elected` event is triggered at some point in the future.

**Comment by arturo-seijas** [comment]
> If I'm understanding this correctly we're relying on the `peer-relation-departed` event happening before `leader-elected`. If it doesn't, we'd remove the nominated unit but not nominate a new unit until another `leader-elected` event is triggered at some point in the future.

Exactly, there should be a leader at any point, so if the celery unit is leader, a leader elected event will be fired when it is removed (if not earlier)

**Comment by mthaddon** [comment]
> Exactly, there should be a leader at any point, so if the celery unit is leader, a leader elected event will be fired when it is removed (if not earlier)

But if the `leader-elected` event is fired before `peer-relation-departed` we'd run the following:
```
if peer_relation and not peer_relation.data[self.app].get("celery-unit"):
```
Since "celery-unit" is set, it would do nothing. Then the `peer-relation-departed` event fires, and we unset this value, and we'd continue with an unset value until the next `leader-elected` event is fired, at some unknown point in the future.

**Comment by arturo-seijas** [comment]
> > Exactly, there should be a leader at any point, so if the celery unit is leader, a leader elected event will be fired when it is removed (if not earlier)
> 
> But if the `leader-elected` event is fired before `peer-relation-departed` we'd run the following:
> 
> ```
> if peer_relation and not peer_relation.data[self.app].get("celery-unit"):
> ```
> 
> Since "celery-unit" is set, it would do nothing. Then the `peer-relation-departed` event fires, and we unset this value, and we'd continue with an unset value until the next `leader-elected` event is fired, at some unknown point in the future.

In that case, the change would occur in the peer relation departed handler. All the units receive that event, so if the celery unit is not the leader, the leader will set that value to itself

**Comment by mthaddon** [comment]
> In that case, the change would occur in the peer relation departed handler. All the units receive that event, so if the celery unit is not the leader, the leader will set that value to itself

Sorry, you are right. Thanks for the out of band explanation, I'd missed that the `peer-relation-departed` event is handling things different if the "celery-unit" value is set to the current leader or not exactly to take care of this possibility.

File: tests/integration/test_charm.py
Comment (jdkandersson) [inline tests/integration/test_charm.py:1]: Do you think we should have a test where the leader is changed? E.g., deploy 2 units, check that celery is working, kill the current leader and wait for the next leader and then check that celery has moved to the new leader unit?

Comment (arturo-seijas) [reply tests/integration/test_charm.py:1]: It seems to me that use case is already covered by the unit tests

**Review by javierdelapuente — APPROVED** [review]

**Review by gregory-schiano — APPROVED** [review]

**Pull request merged**